airbyte-integrations/connectors/source-harvest/source_harvest/streams.py

```diff
@@ -0,0 +1,296 @@
+
+import pendulum
+import requests
+from airbyte_cdk.models import SyncMode
+from airbyte_cdk.sources.streams.http import HttpStream
+
+
+class HarvestStream(HttpStream, ABC):
+    url_base = "https://api.harvestapp.com/v2/"
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        pass
+
+
+class HarvestStreamWithPagination(HarvestStream):
+    per_page = 50
+    primary_key = "id"
+
+    @property
+    def data_field(self) -> str:
+        """
+        :return: Default field name to get data from response
+        """
+        return self.name
+
+    def path(self, **kwargs) -> str:
+        return self.name
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        stream_data = response.json()
+        if stream_data.get("next_page"):
+            return {
+                "page": stream_data["next_page"],
+            }
+
+    def request_params(self, next_page_token: Mapping[str, Any] = None, **kwargs) -> MutableMapping[str, Any]:
+        params = super().request_params(next_page_token=next_page_token, **kwargs)
+        params["per_page"] = self.per_page
+        if next_page_token:
+            params.update(**next_page_token)
+        return params
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        """
+        :return an iterable containing each record in the response
+        """
+        records = response.json().get(self.data_field, [])
+        yield from records
+
+
+class HarvestStreamIncrementalMixin(HttpStream, ABC):
+    cursor_field = "updated_at"
+
+    def __init__(self, updated_since: pendulum.datetime, **kwargs):
+        super().__init__(**kwargs)
+        self._updated_since = updated_since
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
+        """
+        Return the latest state by comparing the cursor value in the latest record with the stream's most recent state object
+        and returning an updated state object.
+        """
+        latest_benchmark = latest_record[self.cursor_field]
+        if current_stream_state.get(self.cursor_field):
+            return {self.cursor_field: max(latest_benchmark, current_stream_state[self.cursor_field])}
+        return {self.cursor_field: latest_benchmark}
+
+    def request_params(self, stream_state: Mapping[str, Any], **kwargs) -> MutableMapping[str, Any]:
+        params = super().request_params(stream_state=stream_state, **kwargs)
+        updated_since = self._updated_since
+        if stream_state.get(self.cursor_field):
+            updated_since = stream_state[self.cursor_field]
+        params.update({"updated_since": updated_since})
+        return params
+
+
+class HarvestStreamWithPaginationSliced(HarvestStreamWithPagination):
+    @property
+    @abstractmethod
+    def parent_stream_name(self) -> str:
```

Comment:
lets have parent class here instead of its name

Reply:
I pass the name because parent classes are incremental streams. But here we need these streams to be not incremental, because otherwise we'll need to pass the `updated_since` value to them (so we could skip some data).